lerobot/common/policies/vqbet/modeling_vqbet.py

```diff
@@ -0,0 +1,2460 @@
+        if len(parents) > 0:
+            parent_module = root_module.get_submodule(".".join(parents))
+        if isinstance(parent_module, nn.Sequential):
+            src_module = parent_module[int(k)]
+        else:
+            src_module = getattr(parent_module, k)
+        tgt_module = func(src_module)
+        if isinstance(parent_module, nn.Sequential):
+            parent_module[int(k)] = tgt_module
+        else:
+            setattr(parent_module, k, tgt_module)
+    # verify that all BN are replaced
+    assert not any(predicate(m) for _, m in root_module.named_modules(remove_duplicate=True))
+    return root_module
+
+
+
+
+class VqVae(nn.Module):
+    def __init__(
+        self, config: VQBeTConfig,
+    ):
+        """
+        VQ-VAE is composed of three parts: encoder, vq_layer, and decoder.
+        Encoder and decoder are MLPs consisting of an input, output layer, and hidden layer, respectively.
+        The vq_layer uses residual VQs.
+        """
+
+        super(VqVae, self).__init__()
+        self.config = config
+
+        self.register_buffer('discretized', torch.tensor(False))
+        self.optimized_steps = 0
+
+        self.vq_layer = ResidualVQ(
+            dim=config.vqvae_embedding_dim,
+            num_quantizers=config.vqvae_groups,
+            codebook_size=config.vqvae_n_embed,
+        )
+
+        self.encoder = MLP(
+            in_channels=self.config.output_shapes["action"][0] * self.config.action_chunk_size,
+            hidden_channels=[config.vqvae_enc_hidden_dim, config.vqvae_enc_hidden_dim, config.vqvae_embedding_dim],
+        )
+        self.decoder = MLP(
+            in_channels=config.vqvae_embedding_dim,
+            hidden_channels=[config.vqvae_enc_hidden_dim, config.vqvae_enc_hidden_dim, self.config.output_shapes["action"][0] * self.config.action_chunk_size],
+        )
+
+        self.train()
+
+    def toggle_discretized(self, state=True):
+        self.discretized = torch.tensor(state)
+
+    def check_discretized(self):
+        return self.discretized.item()
+
+    def eval(self):
+        self.training = False
+        self.vq_layer.eval()
+        self.encoder.eval()
+        self.decoder.eval()
+
+    def train(self, mode=True):
+        """
+        This function forces the RVQ to no longer update when action discretization is complete.
+        Since VQs are partly updated via the EMA method, simply passing data through them can cause unintended modifications.
+        Therefore, we use function overriding to prevent RVQs from being updated during the training of VQ-BeT after discretization completes.
+        """
+        if mode:
+            if self.check_discretized():
+                pass
+            else:
+                self.training = True
+                self.vq_layer.train()
+                self.decoder.train()
+                self.encoder.train()
+        else:
+            self.eval()
+
+    def draw_logits_forward(self, encoding_logits):
```

Comment:
Thank you for your opinion:) I removed `def draw_logits_forward` since it is not used, and changed `def draw_code_forward` to `def get_embeddings_from_code`